File: java-amqp/src/main/java/RPCServer.java
import com.rabbitmq.client.amqp.Connection;
import com.rabbitmq.client.amqp.Environment;
import com.rabbitmq.client.amqp.Management;
import com.rabbitmq.client.amqp.Responder;

import java.nio.charset.StandardCharsets;

public class RPCServer {

    private static final String RPC_QUEUE_NAME = "rpc_queue";

    private static int fib(int n) {
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        return fib(n - 1) + fib(n - 2);
    }

    public static void main(String[] argv) throws Exception {
        Environment environment = TutorialSupport.newEnvironment();
        Connection connection = environment.connectionBuilder().uri(TutorialSupport.BROKER_URI).build();
        try (Management management = connection.management()) {
            management.queue().name(RPC_QUEUE_NAME).quorum().queue().declare();
            management.queuePurge(RPC_QUEUE_NAME);
        }

        System.out.println(" [x] Awaiting RPC requests");

        Responder responder = connection.responderBuilder()
                .requestQueue(RPC_QUEUE_NAME)
                .handler((ctx, req) -> {
                    String response = "";
                    try {
                        String message = new String(req.body(), StandardCharsets.UTF_8);
                        int n = Integer.parseInt(message);
                        System.out.println(" [.] fib(" + message + ")");
                        response += fib(n);
                    } catch (RuntimeException e) {
                        System.out.println(" [.] " + e);
                    }
                    return ctx.message(response.getBytes(StandardCharsets.UTF_8));
                })
                .build();

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            responder.close();
            connection.close();
            environment.close();
        }));

        Thread.sleep(Long.MAX_VALUE);
    }
}

File: go-amqp/rpc_server.go
package main

import (
	"context"
	"log"
	"strconv"

	amqp "github.com/Azure/go-amqp"
	rmq "github.com/rabbitmq/rabbitmq-amqp-go-client/pkg/rabbitmqamqp"
)

const brokerURI = "amqp://guest:guest@localhost:5672/"

func fib(n int) int {
	if n == 0 {
		return 0
	} else if n == 1 {
		return 1
	}
	return fib(n-1) + fib(n-2)
}

func main() {
	ctx := context.Background()
	env := rmq.NewEnvironment(brokerURI, nil)
	conn, err := env.NewConnection(ctx)
	if err != nil {
		log.Panicf("Failed to connect to RabbitMQ: %v", err)
	}
	defer func() {
		_ = env.CloseConnections(context.Background())
	}()

	_, err = conn.Management().DeclareQueue(ctx, &rmq.QuorumQueueSpecification{Name: "rpc_queue"})
	if err != nil {
		log.Panicf("Failed to declare a queue: %v", err)
	}
	_, err = conn.Management().PurgeQueue(ctx, "rpc_queue")
	if err != nil {
		log.Printf("purge: %v", err)
	}

	responder, err := conn.NewResponder(ctx, rmq.ResponderOptions{
		RequestQueue: "rpc_queue",
		Handler: func(_ context.Context, request *amqp.Message) (*amqp.Message, error) {
			var payload []byte
			if len(request.Data) > 0 {
				payload = request.Data[0]
			}
			n, err := strconv.Atoi(string(payload))
			if err != nil {
				return nil, err
			}
			log.Printf(" [.] fib(%d)", n)
			response := fib(n)
			return amqp.NewMessage([]byte(strconv.Itoa(response))), nil
		},
	})
	if err != nil {
		log.Panicf("Failed to create responder: %v", err)
	}
	defer func() { _ = responder.Close(context.Background()) }()

	log.Printf(" [*] Awaiting RPC requests")
	select {}
}
